File: design-patterns/lab3/java/Parrot.java
package hr.fer.zemris.ooup.lab2.model.plugins;

import hr.fer.zemris.ooup.lab2.model.Animal;

public class Parrot extends Animal {
	
	private String name;
	
	public Parrot() {}
	
	public Parrot(String name) {
		this.name = name;
	}
	
	@Override
	public String name() {
		return this.name;
	}

	@Override
	public String greet() {
		return "squawk";
	}

	@Override
	public String menu() {
		return "sjemenke";
	}

}

File: design-patterns/lab3/c/parrot.c
#include <stdio.h>
#include <stdlib.h>
typedef char const* (*PTRFUN)();

typedef struct Parrot{
  PTRFUN* vtable;
  char* name;
} Parrot;

char const* name(void* this) {
    return ((Parrot*)this)->name;
}

char const* greet() {
    return "squawk";
}

char const* menu() {
    return "sjemenke";
}

PTRFUN* vtableParrot[3] = {name, greet, menu};

Parrot* create(char* name) {
    Parrot* p = malloc(sizeof(Parrot));
    construct(p, name);
    return p;
}

void construct(Parrot* p, char* name) {
    p->vtable = vtableParrot;
    p->name = name;
}